// passkit-generator/lib/esm/schemas/index.js
export * from "./Barcode.js";
export * from "./Beacon.js";
export * from "./Location.js";
export * from "./PassFieldContent.js";
export * from "./NFC.js";
export * from "./Semantics.js";
export * from "./PassFields.js";
export * from "./Personalize.js";
export * from "./Certificates.js";
export * from "./UpcomingPassInformation.js";
import Joi from "joi";
import { Barcode } from "./Barcode.js";
import { Location } from "./Location.js";
import { Beacon } from "./Beacon.js";
import { NFC } from "./NFC.js";
import { PassFields } from "./PassFields.js";
import { Semantics } from "./Semantics.js";
import { UpcomingPassInformationEntry } from "./UpcomingPassInformation.js";
import * as Messages from "../messages.js";
import { RGB_HEX_COLOR_REGEX, URL_REGEX } from "./regexps.js";
export const PreferredStyleSchemes = Joi.array().items("posterEventTicket", "eventTicket", 
// or, since iOS 26
"boardingPass", "semanticBoardingPass");
export const RelevantDate = Joi.alternatives(Joi.object().keys({
    startDate: Joi.alternatives(Joi.string().isoDate(), Joi.date().iso()).required(),
    endDate: Joi.alternatives(Joi.string().isoDate(), Joi.date().iso()).required(),
}), Joi.object().keys({
    /**
     * Since iOS 26
     */
    date: Joi.alternatives(Joi.string().isoDate(), Joi.date().iso()),
    /**
     * Since iOS 18, then was renamed in
     * 'date' in iOS 26 (what a breaking change)
     */
    relevantDate: Joi.alternatives(Joi.string().isoDate(), Joi.date().iso()).required(),
}));
export const PassPropsFromMethods = Joi.object({
    nfc: NFC,
    beacons: Joi.array().items(Beacon),
    barcodes: Joi.array().items(Barcode),
    relevantDate: Joi.string().isoDate(),
    relevantDates: Joi.array().items(RelevantDate),
    expirationDate: Joi.string().isoDate(),
    locations: Joi.array().items(Location),
    preferredStyleSchemes: PreferredStyleSchemes,
    upcomingPassInformation: Joi.array().items(UpcomingPassInformationEntry),
});
export const PassKindsProps = Joi.object({
    coupon: PassFields.disallow("transitType"),
    generic: PassFields.disallow("transitType"),
    storeCard: PassFields.disallow("transitType"),
    eventTicket: PassFields.disallow("transitType"),
    boardingPass: PassFields,
});
export const PassType = Joi.string().regex(/(boardingPass|coupon|eventTicket|storeCard|generic)/);
export const OverridablePassProps = Joi.object({
    formatVersion: Joi.number().default(1),
    semantics: Semantics,
    voided: Joi.boolean(),
    logoText: Joi.string(),
    description: Joi.string(),
    serialNumber: Joi.string(),
    appLaunchURL: Joi.string().regex(URL_REGEX),
    teamIdentifier: Joi.string(),
    organizationName: Joi.string(),
    passTypeIdentifier: Joi.string(),
    sharingProhibited: Joi.boolean(),
    groupingIdentifier: Joi.string(),
    suppressStripShine: Joi.boolean(),
    maxDistance: Joi.number().positive(),
    authenticationToken: Joi.string().min(16),
    labelColor: Joi.string().regex(RGB_HEX_COLOR_REGEX),
    stripColor: Joi.string().regex(RGB_HEX_COLOR_REGEX),
    backgroundColor: Joi.string().regex(RGB_HEX_COLOR_REGEX),
    foregroundColor: Joi.string().regex(RGB_HEX_COLOR_REGEX),
    associatedStoreIdentifiers: Joi.array().items(Joi.number()),
    userInfo: Joi.alternatives(Joi.object().unknown(), Joi.array()),
    webServiceURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    bagPolicyURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle posterEventTicket, semanticBoardingPasses
     * @passDomain Event Guide, Semantic Boarding Passes
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    orderFoodURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    parkingInformationURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    directionsInformationURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * URL to a resource to buy or access
     * the parking spot.
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    purchaseParkingURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * URL to a resource to buy the
     * merchandise.
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    merchandiseURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * URL to a resource about public or
     * private transportation to reach the
     * venue.
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    transitInformationURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * URL to a resource about accessibility
     * in the events venue.
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    accessibilityURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * An URL to link experiences to the
     * pass (upgrades and more).
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    addOnURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    contactVenueEmail: Joi.string(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    contactVenuePhoneNumber: Joi.string(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     * @passDomain Event Guide
     *
     * @description
     *
     * To show buttons in the event guide,
     * at least two among those marked with
     * "@passDomain Event Guide" must be used.
     */
    contactVenueWebsite: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * Will add a button among options near "share"
     */
    transferURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * Will add a button among options near "share"
     */
    sellURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * Will remove an automatic shadow in the new
     * event ticket layouts.
     */
    suppressHeaderDarkening: Joi.boolean(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * By default, the chin is colored with a
     * blur. Through this option, it is possible
     * to specify a different and specific color
     * for it.
     */
    footerBackgroundColor: Joi.string().regex(RGB_HEX_COLOR_REGEX),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * Enables the automatic calculation of the
     * `foregroundColor` and `labelColor` based
     * on the background image in the new event
     * ticket passes.
     *
     * If enabled, `foregroundColor` and `labelColor`
     * are ignored.
     */
    useAutomaticColors: Joi.boolean(),
    /**
     * @iOSVersion 18
     * @passStyle eventTicket (new layout)
     *
     * @description
     *
     * Applications AppStore Identifiers
     * related to the event ticket.
     *
     * It is not mandatory for the app to
     * be related to the pass issuer.
     *
     * Such applications won't be able to read
     * the passes users has (probably differently
     * by `associatedStoreIdentifiers`).
     */
    auxiliaryStoreIdentifiers: Joi.array().items(Joi.number()),
    /**
     * @iOSVersion 18.1
     *
     * The text to display next to the logo on posterEventTicket passes.
     */
    eventLogoText: Joi.string(),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for changing the seat for the ticket.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    changeSeatURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for in-flight entertainment.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    entertainmentURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for adding checked bags for the ticket.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    purchaseAdditionalBaggageURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL that links to information to purchase lounge access.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    purchaseLoungeAccessURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for purchasing in-flight wifi.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    purchaseWifiURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for upgrading the flight.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    upgradeURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for management.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    managementURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL for registering a service animal.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    registerServiceAnimalURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL to report a lost bag.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    reportLostBagURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * A URL to request a wheel chair.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    requestWheelchairURL: Joi.string().regex(URL_REGEX),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * The email for the transit provider.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    transitProviderEmail: Joi.string(),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * The phone number for the transit provider.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    transitProviderPhoneNumber: Joi.string(),
    /**
     * @iOSVersion 26
     *
     * @description
     *
     * The URL for the transit provider.
     * Available only with Enhanced (or semantic) Boarding Passes
     */
    transitProviderWebsiteURL: Joi.string().regex(URL_REGEX),
}).with("webServiceURL", "authenticationToken");
export const PassProps = Joi.object()
    .concat(OverridablePassProps)
    .concat(PassKindsProps)
    .concat(PassPropsFromMethods);
export const Template = Joi.object({
    model: Joi.string().required(),
    certificates: Joi.object().required(),
});
// --------- UTILITIES ---------- //
/**
 * Performs validation of a schema on an object.
 * If it fails, will throw an error.
 *
 * @param schema
 * @param data
 */
export function assertValidity(schema, data, customErrorMessage) {
    const validation = schema.validate(data);
    if (validation.error) {
        if (customErrorMessage) {
            console.warn(validation.error);
            throw new TypeError(`${validation.error.name} happened. ${Messages.format(customErrorMessage, validation.error.message)}`);
        }
        throw new TypeError(validation.error.message);
    }
}
/**
 * Performs validation and throws the error if there's one.
 * Otherwise returns a (possibly patched) version of the specified
 * options (it depends on the schema)
 *
 * @param schema
 * @param options
 * @returns
 */
export function validate(schema, options) {
    const validationResult = schema.validate(options, {
        stripUnknown: true,
        abortEarly: true,
    });
    if (validationResult.error) {
        throw validationResult.error;
    }
    return validationResult.value;
}
export function filterValid(schema, source) {
    if (!source) {
        return [];
    }
    return source.reduce((acc, current) => {
        try {
            return [...acc, validate(schema, current)];
        }
        catch (err) {
            console.warn(Messages.format(Messages.FILTER_VALID.INVALID, err));
            return [...acc];
        }
    }, []);
}
